### PR TITLE
Mcloud logger dest

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Logs metrics to MCLI
image

Along for ride:
- minor clean up in wandb test file

# What issue(s) does this change relate to?

[CO-1985](https://mosaicml.atlassian.net/browse/CO-1985)

[CO-1985]: https://mosaicml.atlassian.net/browse/CO-1985?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/loggers/mosaic_logger.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                self.buffered_metadata[key] = format_log_data_value(val)
+            if time.time() - self.time_last_logged > self.log_interval:
+                from mcli.api.exceptions import MAPIException
+                from mcli.sdk import update_run_metadata
+                try:
+                    update_run_metadata(self.run_name, self.buffered_metadata)
+                except MAPIException as e:
+                    log.error(f'Failed to log metadata to Mosaic with error: {e}')
+
+                self.buffered_metadata = {}
```
Comment (aspfohl) [inline composer/loggers/mosaic_logger.py:76]: to prevent data loss?
```suggestion
               else:
                 self.buffered_metadata = {}
```

File: composer/loggers/__init__.py [hunk 2]
```diff
@@ -37,4 +38,5 @@
     'CometMLLogger',
     'ConsoleLogger',
     'MLFlowLogger',
+    'MosaicLogger',
```
Comment (dakinggg) [inline composer/loggers/__init__.py:41]: `MosaicMLLogger`?

File: composer/loggers/mosaic_logger.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+log = logging.getLogger(__name__)
+
+__all__ = ['MosaicLogger']
+
+
+class MosaicLogger(LoggerDestination):
+    """Log to MPlatform.
+
+    Logs metrics to MPlatform. Logging only happens on rank 0 every ``log_interval`` seconds to
```
Comment (dakinggg) [inline composer/loggers/mosaic_logger.py:25]: can logging only on rank 0 be optional? or is that an mcli limitation?

File: composer/loggers/mosaic_logger.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+from composer.loggers.logger import format_log_data_value
+from composer.loggers.logger_destination import LoggerDestination
+from composer.utils import MissingConditionalImportError, dist
+
+log = logging.getLogger(__name__)
+
+__all__ = ['MosaicLogger']
+
+
+class MosaicLogger(LoggerDestination):
```
Comment (dakinggg) [inline composer/loggers/mosaic_logger.py:22]: can you add some flushes of the buffered metadata on like fit end or something? otherwise you'll end up with some data not logged probably

File: composer/loggers/mosaic_logger.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            if time.time() - self.time_last_logged > self.log_interval:
+                from mcli.api.exceptions import MAPIException
+                from mcli.sdk import update_run_metadata
+                try:
+                    update_run_metadata(self.run_name, self.buffered_metadata)
+                except MAPIException as e:
+                    log.error(f'Failed to log metadata to Mosaic with error: {e}')
+
+                self.buffered_metadata = {}
+                self.time_last_logged = time.time()
```
[the remark this answers is not in the document]
Comment (eracah) [reply composer/loggers/mosaic_logger.py:77]: wait we wouldn't want `time_last_logged` to be updated if it failed right?

File: composer/loggers/mosaic_logger.py (hunk 1 above)
Comment (eracah) [inline composer/loggers/mosaic_logger.py:76]: you might want to have the flush logic happen at `batch_end`, `epoch_end`, `fit_end`, etc and only have the adding to buffer happen inside `_log_metadata`. See `ConsoleLogger` code for inspo

File: composer/loggers/mosaic_logger.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            self.buffered_metadata: Dict[str, Any] = {}
+
+            try:
+                import mcli
+                del mcli
+            except ImportError as e:
+                raise MissingConditionalImportError(extra_deps_group='mcli',
+                                                    conda_package='mcli',
+                                                    conda_channel='conda-forge') from e
+            self.run_name = os.environ.get('RUN_NAME')
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/loggers/mosaic_logger.py:50]: I'll add it to the top. Going to skip constants file because we don't have many constants, and it's easier to read everything in a single self contained file, especially if users want to copy paste and add a variant callback

File: composer/loggers/mosaic_logger.py (hunk 3 above)
Comment (mvpatel2000) [reply composer/loggers/mosaic_logger.py:25]: MAPI limitation, otherwise we're going to start overwhelming it

File: composer/loggers/mosaic_logger.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+    def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
+        self._log_metadata(hyperparameters)
+
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self._log_metadata(metrics)
+
+    def _log_metadata(self, metadata: Dict[str, Any]) -> None:
+        if self._enabled:
+            for key, val in metadata.items():
+                self.buffered_metadata[key] = format_log_data_value(val)
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/loggers/mosaic_logger.py:67]: Tensors are not JSON serializable unfortunately. I can change it to keep in primitive types. For tensors which are not singletons, eg loss, we do 
>         return 'Tensor of shape ' + str(data.shape)

Does that work? any suggestions?

File: composer/loggers/mosaicml_logger.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+                            'is disabled as it is unable to identify which run to log to.')
+                self._enabled = False
+
+    def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
+        self._log_metadata(hyperparameters)
+
+    def log_metrics(self, metrics: Dict[str, Any], step: Optional[int] = None) -> None:
+        self._log_metadata(metrics)
+
+    def batch_end(self, state: State, logger: Logger) -> None:
```
Comment (dakinggg) [inline composer/loggers/mosaicml_logger.py:74]: are you sure you want to flush on every batch? that doesn't limit the sending _that_ much

File: composer/loggers/mosaicml_logger.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    from composer.core import State
+
+log = logging.getLogger(__name__)
+
+__all__ = ['MosaicMLLogger']
+
+RUN_NAME_ENV_VAR = 'RUN_NAME'
+
+
+class MosaicMLLogger(LoggerDestination):
```
Comment (dakinggg) [inline composer/loggers/mosaicml_logger.py:32]: should we have a `filter_keys` or something (i.e. if you want to avoid sending optimizer monitor logs to mcli or something)

File: composer/loggers/mosaicml_logger.py (hunk 8 above)
Comment (mvpatel2000) [reply composer/loggers/mosaicml_logger.py:74]: It calls the flush function, but it won't flush unless the interval has passed

File: composer/loggers/mosaicml_logger.py (hunk 9 above)
Comment (mvpatel2000) [reply composer/loggers/mosaicml_logger.py:32]: Good call... yea let me add something + unit test it

File: composer/loggers/mosaicml_logger.py (hunk 8 above)
Comment (dakinggg) [reply composer/loggers/mosaicml_logger.py:74]: oh right I kinda missed that. the intention of the flush is to _write even if the interval has not passed_. thats why  suggested running it at fit end (and maybe some other common checkpoints). Also would be good to flush on a crash if possible (i.e. in the close or post_close or whatever).

**Review by eracah — APPROVED** [review]
Nice. All my feedback was addressed. LGTM

**Comment by eracah** [comment]
> Nice. All my feedback was addressed. LGTM

Oh wait. Tests?

File: composer/loggers/mosaic_logger.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/mosaic_logger.py]: lets just stick with the current thing of just printing the shape?

File: tests/loggers/test_wandb_logger.py [hunk 10]
```diff
@@ -15,12 +15,8 @@
 import torch
 from torch.utils.data import DataLoader
```
Comment (dakinggg) [inline tests/loggers/test_wandb_logger.py:16]: Were the changes in this file cleanup?

**Review by dakinggg — APPROVED** [review]
Approving to unblock, but would like some tests

File: tests/loggers/test_wandb_logger.py (hunk 10 above)
Comment (mvpatel2000) [reply tests/loggers/test_wandb_logger.py:16]: Yes, just cleaning up some imports

File: composer/trainer/trainer.py [hunk 11]
```diff
@@ -824,7 +824,7 @@ def __init__(
         # Save Checkpoint
         save_folder: Optional[str] = None,
         save_filename: str = 'ep{epoch}-ba{batch}-rank{rank}.pt',
-        save_latest_filename: Optional[str] = 'latest-rank{rank}.pt',
+        save_latest_filename: str = 'latest-rank{rank}.pt',
```
Comment (dakinggg) [inline composer/trainer/trainer.py:827]: Wait, isn't this actually optional? If none, won't save latest, right?

Comment (mvpatel2000) [reply composer/trainer/trainer.py]: oops... good save...

**Review by bandish-shah — APPROVED** [review]
LGTM

File: setup.py [hunk 12]
```diff
@@ -206,7 +206,13 @@ def package_files(prefix: str, directory: str, extension: str):
     'onnxruntime>=1.12.1,<2',
 ]
 
-extra_deps['mlflow'] = ['mlflow>=2.0.1,<3.0']
+extra_deps['mlflow'] = [
+    'mlflow>=2.0.1,<3.0',
+]
+
+extra_deps['mcli'] = [
+    'mosaicml-cli>=0.3.6,<0.4',
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply setup.py:214]: I think there's a few semantic changes at 0.4.0, and I'd rather pin to be safe. I'll loosen after that change set and possibly make this a required dep since mcli will become a lot lighter.... I'd rather be safe for now
